[WIP] Integrate AliTPCCommon into O2 + merge new features

davidrohr opened the pull request:
Integrates AliTPCCommon (or now called GPU) into O2

- Merges AliTPCCommon into O2
- Renames AliTPCCommon to GPU
- Includes all AliTPCCommon patches between 2.3.2.1 and 2.5 + related O2 patches, which could not be merged before due to problems in AliRoot.
- Contains new ITS GPU track fit jointly developed in both repositories.

ping @ehellbar @matthiasrichter @sawenzel @dberzano @martenole

matthiasrichter commented:
Great effort, thanks!

Given the extent of this PR I think we should define a list of features which are supposed to work, and check if we have a functional unit test for each of them.

matthiasrichter commented:
I have tested the PR with the `tpc-reco-workflow`. The number of found tracks is identical, but the parameters of the tracks seem to be different, e.g. the time0 parameter.

Reference:
![tpc-reco-workflow_reference](https://user-images.githubusercontent.com/8779452/53625339-1d8f2680-3c03-11e9-9c7b-ca70e03b908e.png)

With this PR
![tpc-reco-workflow_pr1747](https://user-images.githubusercontent.com/8779452/53625349-2253da80-3c03-11e9-835a-a8cc4a6af025.png)

davidrohr commented:
> Great effort, thanks!
> 
> Given the extent of this PR I think we should define a list of features which are supposed to work, and check if we have a functional unit test for each of them.

Indeed, I'll add some tests using a gun particle to test features such as merging accross sectors, accross central electrode, and following into adjacent sectors.
We can test the GPU workflow by running the GPU memory allocation scheme on the CPU version and comparing to the normal CPU version.

For ITS / TRD tracking / transformation / SCD it will by up to @mpuccio @martenole @sgorbuno @ehellbar to provide unit tests.

davidrohr commented:
> I have tested the PR with the `tpc-reco-workflow`. The number of found tracks is almost identical, but the parameters of the tracks seem to be completely different, e.g. the time0 parameter.
> 
> 
Hi Matthias, this looks like the first test was in continuous mode and the second in triggered mode. Are you sure you ran the same? Perhaps the flag from the initialization is not properly propagated. I'll check.

mpuccio commented:
Hi, on the unit test: if a particle gun is available we can test directly the our tracker there. As already discussed in several occasions, it would be also good to have some central production of events that monitors on a larger scale (every week, let say) the performance (both computing time and physics) of simulation and reconstruction.

davidrohr commented:
@mpuccio : I think we'd do two things: we should have a moderate simulation with full reco and eff/res qa as a global tests, but we can use some gun particles to test individual features. Right now for the TPC, I am doing this just manually without any distortions etc. and I am just using a straight line.
Using a gun in o2sim is probably complicated because it would mean we have to run the full chain with o2sim / digitization / tracking for each unit test.

davidrohr commented:
@matthiasrichter @sawenzel @dberzano : II think the ctest tests all ROOT macros it found. Since we included also the AliRoot part of AliTPCCommon, some macros fail because of missing AliRoot headers (AliMagF.h etc.) We cannot fix this. What would be the best way to remove these macros from the test?

shahor02 commented on GPU/Common/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+    AliGPUCommonTransform3D.h
+	FlatObject.h
+)
+
```
@davidrohr could you add a LinkDef file with at least 
``#pragma link C++ class ali_tpc_common::Base::FlatObject+;``, to allow root persistency for derived objects.
BTW, should not the namespace become gpu_common?

davidrohr commented:
@shahor02 : I have the patch with the linkdef already here, but I wanted to make sure the old version compiles before I add more stuff. It'll be in the final version of the PR for sure.
All the Ali... ali_... will also go.

shahor02 commented:
> Using a gun in o2sim is probably complicated because it would mean we have to run the full chain with o2sim / digitization / tracking for each unit test

@sawenzel knows how to indicate the dependencies of the test and use the output of one test in the dependent once.

davidrohr commented:
this will take a long time.... ROOT5 CINT segfaults with the new version...

matthiasrichter commented:
> > I have tested the PR with the `tpc-reco-workflow`. The number of found tracks is almost identical, but the parameters of the tracks seem to be completely different, e.g. the time0 parameter.
> 
> Hi Matthias, this looks like the first test was in continuous mode and the second in triggered mode. Are you sure you ran the same? Perhaps the flag from the initialization is not properly propagated. I'll check.

I will have a look, maybe have been too sloppy with this flag, very likly I have to change this in the workflow specs.

davidrohr commented:
When the time0 is different, this will make all parameters slightly different, as we'll use different z for inhomogeneous b-field and cluster error parameterization.
Anyhow, AliTPCCommon 2.4 and 2.5 enabled more track fit features which were disabled as they were unstable in 2.3.2, therefore the parameters will be different in any case.
For checking only before and after the merge, you'll have to compare the to former 2.5 tag...
I guess when my general QA looks sane, I'll believe everything is all right.

davidrohr commented:
Hi all,
so the status is:
- All code is merged, including all new recent developments like ITS track fit.
- Code compiles with AliRoot / O2 / Standalone with ROOT5 / ROOT6 for CPU / CUDA / HIP / OpenCL.
- Verified AliRoot reconstruction and O2 simulation with TPC workflow.
- I applied clang-format to all files, and fixed the braces and member variable names to match the coding guidelines. Fixed some doxygen warnings along this. More work is needed for function calls and local variables, but I guess this does not have to happen immediately.
- Squashed former AliRoot commits in the history which were not related to GPU/TPC tracking.
- All codacy and codechecker issues in the GPU tracking code are fixed, except for false alarms and modernizations requests that are incompatible to OpenCL or ROOT 5. Theoretically, one could invent workaround using plenty of preprocessor defines, but I'd like to avoid making the code more unreadable just to get rid of the warnings, so I'd ignore them for now.

What still needs to be done:
- Fix copyright notices in TPCSpaceCharge and TRDTracking code.
- Fix audacity warnings in TPCSpaceCharge and TRDTracking code.
- Fix codechecker issues in TPCSpaceCharge code (I already went over the TRD code and applied everything that does not collide with the GPU).
- Fix function / local variable names to comply to coding guidelines and go through doxygen warnings (in general)
@ehellbar @martenole : Could you have a look for the SpaceCharge and TRD code.

I'll be busy with other things next week, so I cannot go on here.
In my opinion, we can merge this PR as is, and clean up the rest over time? @sawenzel @matthiasrichter 
When this one is merged, I'll update the corresponding PR to AliRoot alisw/AliRoot#914 and remove the alitpccommon recipe.

davidrohr commented:
Before we merge this: @matthiasrichter : Did you check the tpc workflow for continuous data? I'll also have a look at it. Also, I figured some of the files I took from AliRoot/HLT have an incompatible license statement. I am discussing with Volker to change it to O2 lincense, but I am not entitled to remove the license statement by myself.

davidrohr commented:
@matthiasrichter : I can confirm there is indeed a bug in the continuous tracking. Looking into it.

ktf commented:
I suspect this PR is polluting the CI cache so that CMake incremental builds do not work. I am closing this to verify this is actually the case.

sawenzel commented:
The files will probably still be present in the build dir?

ktf commented:
No, that will not happen because the new container will use a new build directory. I think I saw that some PR are back into green. I think we are actually hitting a cmake / make design limitation when files are moved.

davidrohr commented:
Indeed, this is also what I suspected. Although I am wondering why cmake is taking care of the incremental build correctly. But it shouldn't happen that often, so perhaps we don't have to bother with it too much.

I can finalize the fix for the continuous mode and this license business offline, and then we can reopen the PR just before merging it. This way we won't disturb others.
